src/shell/shell_misc.cpp

```diff
@@ -20,479 +20,421 @@
 #include "shell.h"
 
 #include <algorithm>
-#include <cstddef>
-#include <cstdio>
-#include <cstdlib>
 #include <cstring>
-#include <iterator>
-#include <regex>
 
-#include "regs.h"
+#include "../ints/int10.h"
 #include "callback.h"
+#include "regs.h"
 #include "string_utils.h"
-#include "../ints/int10.h"
 
-DOS_Shell::~DOS_Shell() {
+namespace {
+[[nodiscard]] std::vector<std::string> get_completions(std::string_view command);
+void run_binary_executable(std::string_view fullname, std::string_view args);
+
+class Int10Manager {
+public:
+	Int10Manager();
+	void Update(std::string_view command, std::string::size_type cursor);
+	void Newline();
+
+	[[nodiscard]] static uint32_t MaxCommandSize();
+
+private:
+	struct CursorPosition {
+		uint8_t page   = 0;
+		uint8_t row    = 0;
+		uint8_t column = 0;
+	};
+	CursorPosition position_zero = {};
+
+	std::string::size_type previous_size = 0;
+
+	static void outc(uint8_t c);
+	void SetCurrentPositionToZero();
+	void SetCursor(std::string::size_type index);
+};
+} // namespace
+
+DOS_Shell::~DOS_Shell()
+{
 	bf.reset();
 }
 
-void DOS_Shell::ShowPrompt(void) {
-	uint8_t drive=DOS_GetDefaultDrive()+'A';
+void DOS_Shell::ShowPrompt()
+{
+	uint8_t drive = DOS_GetDefaultDrive() + 'A';
 	char dir[DOS_PATHLENGTH];
-	reset_str(dir); // DOS_GetCurrentDir doesn't always return
-	                // something. (if drive is messed up)
+	reset_str(dir);
+
+	// DOS_GetCurrentDir doesn't always return something.
+	// (if drive is messed up)
 	DOS_GetCurrentDir(0, dir);
 	InjectMissingNewline();
 	WriteOut("%c:\\%s>", drive, dir);
-	ResetLastWrittenChar('\n'); // prevents excessive newline if cmd prints nothing
-}
 
-static void outc(uint8_t c) {
-	uint16_t n=1;
-	DOS_WriteFile(STDOUT,&c,&n);
+	// prevents excessive newline if cmd prints nothing
+	ResetLastWrittenChar('\n');
 }
 
-static void move_cursor_back_one() {
-	const uint8_t page = real_readb(BIOSMEM_SEG, BIOSMEM_CURRENT_PAGE);
-	if (CURSOR_POS_COL(page) > 0) {
-		outc(8);
-	} else if (const auto row = CURSOR_POS_ROW(page); row > 0) {
-		BIOS_NCOLS;
-		INT10_SetCursorPos(row - 1, ncols - 1, page);
+void DOS_Shell::InputCommand(char* line)
+{
+	std::string command = ReadCommand();
+	history.emplace_back(command);
+
+	const auto* const dos_section = dynamic_cast<Section_prop*>(
+	        control->GetSection("dos"));
+	assert(dos_section != nullptr);
+	const std::string_view expand_shell_variable_pref = dos_section->Get_string(
+	        "expand_shell_variable");
+	if ((expand_shell_variable_pref == "auto" && dos.version.major >= 7) ||
+	    expand_shell_variable_pref == "true") {
```

Comment:
@johnnovak , I remember the improved config `bool` handling (where "1", "on", "true", and "yes" all are treated as positive and the opposites for negative).

I'm looking at:

```c++
bool Value::SetBool(const std::string& in)
{
	istringstream input(in);
	string result;
	input >> result;
	lowcase(result);
	_bool = false;

	if (!result.size()) {
		return false;
	}

	if (result == "0" || result == "disabled" || result == "false" ||
	    result == "off" || result == "no") {
		_bool = false;
	} else if (result == "1" || result == "enabled" || result == "true" ||
	           result == "on" || result == "yes") {
		_bool = true;
	} else {
		return false;
	}

	return true;
}
```

Ideally we could leverage that more robust bool handling here too (even though this is a conf string type instead of conf bool.

Reply:
Yeah I guess finding & extracting my function that parses the config bools is in order then into string utils or something 😄 Good that you remembered this! 🎉

Reply:
Are you planning to do this @MeAreJeenius ?

Reply:
Let's go w/ a separate refactoring and get this PR inbound!